### PR TITLE
MRESOLVER-116 document configuration options

### PR DESCRIPTION
**Pull request by kwin**

### CODE REVIEW

**Comment by kwin** [comment]
Help appreciated for the configuration properties marked with `?`

1. `aether.connector.smartChecksums`
1. `aether.artifactResolver.snapshotNormalization`
1. `aether.updateCheckManager.sessionState`

**Comment by michael-o** [comment]
This is up next after all other assigned tickets.

File: src/site/markdown/configuration.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+`aether.checksums.forSignature` | Boolean | Flag indicating if signature artifacts (`.asc`) should have checksums. | `false` | no
+`aether.checksums.algorithms` | String | List of [algorithms](https://docs.oracle.com/en/java/javase/11/docs/specs/security/standard-names.html#messagedigest-algorithms) passed to [MessageDigest](https://docs.oracle.com/javase/8/docs/api/java/security/MessageDigest.html). for which checksums are validated (downloaded) and generated (uploaded). | `"SHA-512,SHA-256,SHA-1,MD5"` | no
+`aether.dependencyCollector.maxExceptions` | Integer | Only exceptions up to the number given in this configuration property are emitted. Exceptions which exceed that number are swallowed. | `50` | no
+`aether.dependencyCollector.maxCycles` | Integer | Only up to the given amount cyclic dependencies are emitted. | `10` | no
+`aether.artifactResolver.snapshotNormalization` | Boolean | ? | `true` | no
+`aether.metadataResolver.threads` | Integer | Number of threads to use in parallel for resolving metadata. | `4`| no
```
Comment (michael-o) [inline src/site/markdown/configuration.md:33]: Space missing after `4`.

File: src/site/markdown/configuration.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+`aether.enhancedLocalRepository.trackingFilename` | String | Filename of the file in which to track the remote repositories. | `"_remote.repositories"` | no
+`aether.offline.protocols` | String | Comma-separated list of protocols which are supposed to be resolved offline. | - | no
+`aether.offline.hosts` | String | Comma-separated list of hosts which are supposed to be resolved offline. | - | no
+`aether.updateCheckManager.sessionState` | String | ? | `true` | no
+
+All properties which have `yes` in the column `Supports Repo ID Suffix` can be optionally configured specifically for a repository id. In that case the configuration property needs to be suffixed with the repository id of the repository to configure.
```
Comment (michael-o) [inline src/site/markdown/configuration.md:39]: I think this should provide a config how such a property name should look like.

**Comment by michael-o** [comment]
As far as I understand the code for `aether.connector.smartChecksums` it instructs Resolver to validate the checksums from remote with the checksums calculated by Resolver.

**Comment by michael-o** [comment]
`aether.artifactResolver.snapshotNormalization` is obvious in the code: It replaces the the timestamped snapshot file name with a filename containing `SNAPSHOT` only.

**Comment by michael-o** [comment]
`sessionState` is used in `DefaultUpdateCheckManager`. Have a look at the code.

**Comment by michael-o** [comment]
How did you determine the list of properties?

At least these are missing:
* `aether.connector.basic.threads`
* `aether.dependencyManager.verbose`

File: src/site/markdown/configuration.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+`aether.connector.userAgent` | String | The user agent that repository connectors should report to servers. |  `"Aether"` | no
+`aether.connector.resumeDownloads` | Boolean | Whether to resume partially downloaded files if the download has been interrupted. | `true` | yes
+`aether.connector.resumeThreshold` | Long | The size in bytes which a partial download needs to have at least to be resumed. Requires `aether.connector.resumeDownloads` to be `true` to be effective. | `64 * 1024` | yes
+`aether.connector.requestTimeout` | Long | Request timeout in milliseconds. | `1800000` | yes
+`aether.connector.connectTimeout` | Long | Connect timeout in milliseconds. | `10000` | yes
+`aether.connector.wagon.config` | Object | The configuration to use for the wagon provider. | - | yes (must be used)
+`aether.connector.http.headers` | String | The request headers to use for HTTP-based repository connectors. The headers are specified using a map of Strings mapping a header name to its value. The repository-specific headers map is supposed to be complete, i.e. is not merged with the general headers map. | - | yes
```
Comment (michael-o) [inline src/site/markdown/configuration.md:17]: The type is incorrect. I have found `ConfigUtils.getMap()` with this property.

**Comment by kwin** [comment]
> Please use a consistent Java version which we support as a minimum.

I don't get this remark as this is only about a documentation update. I changed nothing about java versions. Please elaborate.

File: src/site/markdown/configuration.md (hunk 3 above)
Comment (kwin) [reply src/site/markdown/configuration.md:17]: You are right, IMHO those are configurable from Maven either. Will create a dedicated bug report to make both Map<> and List<> work with the system properties used by Maven.

**Comment by michael-o** [comment]
> 
> 
> > Please use a consistent Java version which we support as a minimum.
> 
> I don't get this remark as this is only about a documentation update. I changed nothing about java versions. Please elaborate.

The Java security links refer to Java 11 which we don't use/require. Our minimum version is 7. Therefore, the links shall use this one.

**Comment by kwin** [comment]
@michael-o I checked once more all occurrences of `ConfigUtils` and I think now I collected all properties and also I have integrated your feedback. Please check again.

File: src/site/markdown/configuration.md [hunk 4]
```diff
@@ -2,41 +2,42 @@
...
 `aether.connector.classpath.loader` | ClassLoader | `ClassLoader` from which resources should be retrieved which start with the `classpath:` protocol. | `Thread.currentThread().getContextClassLoader()` | no
 `aether.checksums.forSignature` | Boolean | Flag indicating if signature artifacts (`.asc`) should have checksums. | `false` | no
-`aether.checksums.algorithms` | String | List of [algorithms](https://docs.oracle.com/en/java/javase/11/docs/specs/security/standard-names.html#messagedigest-algorithms) passed to [MessageDigest](https://docs.oracle.com/javase/8/docs/api/java/security/MessageDigest.html). for which checksums are validated (downloaded) and generated (uploaded). | `"SHA-512,SHA-256,SHA-1,MD5"` | no
+`aether.checksums.algorithms` | String | List of [algorithms](https://docs.oracle.com/javase/7/docs/technotes/guides/security/StandardNames.html#MessageDigest) passed to [MessageDigest](hhttps://docs.oracle.com/javase/7/docs/api/java/security/MessageDigest.html) with which checksums are validated (downloaded) and generated (uploaded). | `"SHA-512,SHA-256,SHA-1,MD5"` | no
```
Comment (michael-o) [inline src/site/markdown/configuration.md:29]: hhttps => hhttps

File: src/site/markdown/configuration.md [hunk 5]
```diff
@@ -2,41 +2,42 @@
...
 `aether.dependencyCollector.maxExceptions` | Integer | Only exceptions up to the number given in this configuration property are emitted. Exceptions which exceed that number are swallowed. | `50` | no
 `aether.dependencyCollector.maxCycles` | Integer | Only up to the given amount cyclic dependencies are emitted. | `10` | no
-`aether.artifactResolver.snapshotNormalization` | Boolean | ? | `true` | no
-`aether.metadataResolver.threads` | Integer | Number of threads to use in parallel for resolving metadata. | `4`| no
+`aether.dependencyManager.verbose` | Boolean | A flag controlling the verbose mode for dependency management. If enabled, the original attributes of a dependency before its update due to dependency managemnent will be recorded in the node's `DependencyNode#getData()` when building a dependency graph. | `false` | no
+`aether.artifactResolver.snapshotNormalization` | Boolean | It replaces the the timestamped snapshot file name with a filename containing SNAPSHOT only. This only affects resolving/retrieving artifacts but not uploading those. | `true` | no
```
Comment (michael-o) [inline src/site/markdown/configuration.md:33]: > `SNAPSHOT` qualifier only.

File: src/site/markdown/configuration.md [hunk 6]
```diff
@@ -2,41 +2,42 @@
...
+`aether.metadataResolver.threads` | Integer | Number of threads to use in parallel for resolving metadata. | `4` | no
 `aether.enhancedLocalRepository.trackingFilename` | String | Filename of the file in which to track the remote repositories. | `"_remote.repositories"` | no
 `aether.offline.protocols` | String | Comma-separated list of protocols which are supposed to be resolved offline. | - | no
 `aether.offline.hosts` | String | Comma-separated list of hosts which are supposed to be resolved offline. | - | no
-`aether.updateCheckManager.sessionState` | String | ? | `true` | no
+`aether.updateCheckManager.sessionState` | String | Manages the session state, i.e. influences if the same download requests to artifacts/metadata will happen multiple times within the same RepositorySystemSession. If `"true"` will enable the session state. If `"bypass"` will enable bypassing (i.e. store all artifact ids/metadata ids which have been updates but not evaluating those). All other values lead to disabling the session state completely. | `"true"` | no
```
Comment (michael-o) [inline src/site/markdown/configuration.md:38]: Don't quote boolean as you didn't quote them throughout.

File: src/site/markdown/configuration.md [hunk 7]
```diff
@@ -2,41 +2,42 @@
...
 
-All properties which have `yes` in the column `Supports Repo ID Suffix` can be optionally configured specifically for a repository id. In that case the configuration property needs to be suffixed with the repository id of the repository to configure.
+All properties which have `yes` in the column `Supports Repo ID Suffix` can be optionally configured specifically for a repository id. In that case the configuration property needs to be suffixed with a comma followed by the repository id of the repository to configure, e.g. `aether.connector.http.headers.central` for repository with id `central`.
```
Comment (michael-o) [inline src/site/markdown/configuration.md:40]: You obviously meant period and not comma, didn't you?

File: src/site/markdown/configuration.md [hunk 8]
```diff
@@ -2,41 +2,42 @@
...
+`aether.connector.perms.dirMode` | String | [Octal numerical notation of permissions](https://en.wikipedia.org/wiki/File_system_permissions#Numeric_notation) to set for newly created directories. Only considered by certain Wagon providers. | - | no
 `aether.connector.perms.group` | String | The group which should own newly created directories/artifacts. Only considered by certain wagon providers. | - | no
 `aether.connector.persistedChecksums` | Boolean | A flag indicating whether checksums which are retrieved during checksum validation should be persisted in the local filesystem next to the file they provide the checksum for. | `true` | no
-`aether.connector.smartChecksums` | Boolean | ? | `true` | no
+`aether.connector.smartChecksums` | Boolean | A flag indicating that instead of comparing the explicit checksum from the remote repo with the calculated one it will try to extract it from the actual artifact requests's response header named `ETag` in format `SHA1{<checksum>}`. This only works for HTTP(S) requests and certain transport extensions. In addition it only supports SHA1. | `true` | no
```
Comment (michael-o) [inline src/site/markdown/configuration.md:26]: I will look at this one again.

Comment (michael-o) [reply src/site/markdown/configuration.md:26]: * This is almost correct, but the format is for Nexus 2.14.18: `"{SHA1{<checksum>}}"`.
* Correct SHA1 to SHA-1 in the last sentence.

File: src/site/markdown/configuration.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+`aether.metadataResolver.threads` | Integer | Number of threads to use in parallel for resolving metadata. | `4` | no
+`aether.enhancedLocalRepository.trackingFilename` | String | Filename of the file in which to track the remote repositories. | `"_remote.repositories"` | no
+`aether.offline.protocols` | String | Comma-separated list of protocols which are supposed to be resolved offline. | - | no
+`aether.offline.hosts` | String | Comma-separated list of hosts which are supposed to be resolved offline. | - | no
+`aether.updateCheckManager.sessionState` | String | Manages the session state, i.e. influences if the same download requests to artifacts/metadata will happen multiple times within the same RepositorySystemSession. If `"true"` will enable the session state. If `"bypass"` will enable bypassing (i.e. store all artifact ids/metadata ids which have been updates but not evaluating those). All other values lead to disabling the session state completely. | `"true"` | no
```
Comment (michael-o) [inline src/site/markdown/configuration.md:38]: This in an interesting one. I think `true` makes no sense here. It should be `enabled`. I will spawn a ticket to support this too.

File: src/site/markdown/configuration.md (hunk 6 above)
Comment (kwin) [reply src/site/markdown/configuration.md:38]: It really is a string here!

File: src/site/markdown/configuration.md (hunk 7 above)
Comment (kwin) [reply src/site/markdown/configuration.md:40]: Indeed.

File: src/site/markdown/configuration.md [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+## Property Type Conversion
+
+If the value is not given in the target type the following conversions are applied.
+
+From | To | With
+--- | --- | ---
+`String` | `boolean` | [`Boolean.parseBoolean(...)`](https://docs.oracle.com/javase/7/docs/api/java/lang/Boolean.html#parseBoolean(java.lang.String))
+`String` | `int` | [`Integer.valueOf(...)`](https://docs.oracle.com/javase/7/docs/api/java/lang/Integer.html#valueOf(java.lang.String))
+`String` | `long` | [`Long.valueOf(...)`](https://docs.oracle.com/javase/7/docs/api/java/lang/Long.html#valueOf(java.lang.String))
+`String` | `float` | [`Float.valueOf(...)`](https://docs.oracle.com/javase/7/docs/api/java/lang/Float.html#valueOf(java.lang.String))
```
Comment (michael-o) [inline src/site/markdown/configuration.md:51]: This does not correspond to #61 anymore.

File: src/site/markdown/configuration.md (hunk 9 above)
Comment (michael-o) [reply src/site/markdown/configuration.md:38]: Created #62

**Comment by jira-importer** [comment]
Resolve #869

**Comment by jira-importer** [comment]
Resolve #869

**Pull request closed without merging**